beacon_node/beacon_chain/src/migrate.rs

```diff
@@ -504,249 +476,278 @@ impl<E: EthSpec, Hot: ItemStore<E>, Cold: ItemStore<E>> BackgroundMigrator<E, Ho
+                // Note: The sanity check above for existance of at least one summary with
+                // new_finalized_checkpoint.root should ensure that this call never errors
+                block_summaries_dag
+                    .descendant_block_roots_of(&new_finalized_checkpoint.root)
+                    .map_err(PruningError::SummariesDagError)?,
+            ),
+        );
 
-            if let Some(abandoned_head) = potentially_abandoned_head {
-                debug!(
-                    log,
-                    "Pruning head";
-                    "head_block_root" => format!("{:?}", abandoned_head),
-                    "head_slot" => head_slot,
-                );
-                abandoned_heads.insert(abandoned_head);
-                abandoned_blocks.extend(
-                    potentially_abandoned_blocks
-                        .iter()
-                        .filter_map(|(_, maybe_block_hash, _)| *maybe_block_hash),
-                );
-                abandoned_states.extend(potentially_abandoned_blocks.iter().filter_map(
-                    |(slot, _, maybe_state_hash)| maybe_state_hash.map(|sr| (*slot, sr)),
-                ));
+        // Note: ancestors_of includes the finalized state root
+        let newly_finalized_state_summaries = state_summaries_dag
+            .ancestors_of(new_finalized_state_hash)
+            .map_err(PruningError::SummariesDagError)?;
+        let newly_finalized_state_roots = newly_finalized_state_summaries
+            .iter()
+            .map(|(root, _)| *root)
+            .collect::<HashSet<Hash256>>();
+        let newly_finalized_states_min_slot = *newly_finalized_state_summaries
+            .iter()
+            .map(|(_, slot)| slot)
+            .min()
+            .ok_or(PruningError::EmptyFinalizedStates)?;
+
+        // Note: ancestors_of includes the finalized block
+        let newly_finalized_blocks = block_summaries_dag
+            .ancestors_of(new_finalized_checkpoint.root)
+            .map_err(PruningError::SummariesDagError)?;
+        let newly_finalized_block_roots = newly_finalized_blocks
+            .iter()
+            .map(|(root, _)| *root)
+            .collect::<HashSet<Hash256>>();
+        let newly_finalized_blocks_min_slot = *newly_finalized_blocks
+            .iter()
+            .map(|(_, slot)| slot)
+            .min()
+            .ok_or(PruningError::EmptyFinalizedBlocks)?;
+
+        // We don't know which blocks are shared among abandoned chains, so we buffer and delete
+        // everything in one fell swoop.
+        let mut blocks_to_prune: HashSet<Hash256> = HashSet::new();
+        let mut states_to_prune: HashSet<(Slot, Hash256)> = HashSet::new();
+
+        for (slot, summaries) in state_summaries_dag.summaries_by_slot_ascending() {
+            for (state_root, summary) in summaries {
+                let should_prune =
+                    if finalized_and_descendant_block_roots.contains(&summary.latest_block_root) {
```

Comment:
Note to self: I think we should prune based on ancestors of the finalized _state_. This will more neatly handle the case where we have e.g.

- block at slot 30 (canonical)
- block at slot 31 building on 30 (not canonical)
- skipped slot at slot 32 that is finalized (also finalizes slot 31 as canonical)

In this example, the block at slot 31 descends from the finalized block but not from the finalized checkpoint. So it should be pruned, along with all its descendants.